src/dodal/common/visit.py

```diff
@@ -150,3 +151,42 @@ def __call__(self, device_name: str | None = None) -> PathInfo:
         return PathInfo(
             directory_path=self._root, filename=self._filename_provider(device_name)
         )
+
+
+DEFAULT_TEMPLATE = "{device_name}-{instrument}-{scan_id}"
+
+
+class StartDocumentBasedPathProvider(PathProvider):
+    """A PathProvider that sources from metadata in a RunStart document.
+
+    This uses metadata a RunStart document to determine file names and data session
+    directories. The file naming defaults to "{device_name}-{instrument}-{scan_id}", so
+    the file name is incremented by scan number. A template can be included in the
+    StartDocument to allow for custom naming conventions.
+
+    """
+
+    def __init__(self) -> None:
+        self._doc = {}
+
+    def update_run(self, name: str, start_doc: RunStart) -> None:
+        """Cache a start document.
+
+        This can be plugged into the run engines subscribe method.
```

Comment:
The RunEngine subscribe method can be passed the names of documents you expect to receive, if the name check is just for sanity's sake it's fine, but if you're receiving all of the Document types because you're not making use of it, I would add it in, else the start_doc param may cause issues when passed another Document type.

RE.subscribe(update_run, name="start")

Reply:
This is done in blueapi, so we are only getting start documents.

Reply:
Maybe nitpicky but this looks weird to me because all MX callbacks inherit from `CallbackBase` and would use `def start(` for this. I appreciate that's not easy given that `PathProvider` and `CallbackBase` both want to make use of `__call__` though. I think it's particularly confusing because the definition and the actual usecase is split across different repos

Reply:
We made an issue for this refactor here. https://github.com/DiamondLightSource/dodal/issues/1132

Reply:
It was a suggestion, happy to leave as is